add unit tests for cmd/bundle

jberkhahn opened the pull request:
**Description of the change:**
Add unit tests for the `bundle validate` and `bundle create` commands.

**Motivation for the change:**
Golang unit tests are another layer of robust CI when paired with the e2e and script testing we currently have.

jberkhahn commented:
@estroz started work on this. I'm leaving the Run() tests as pending for the moment as I'm not sure how to test that stuff without doing the counterfeiter refactor, which we should probably wait for the KB integration before we do that. Please tag anyone else you think might be interested in this.

jberkhahn commented:
this should be good to go

camilamacedo86 commented on cmd/operator-sdk/bundle/validate_test.go:
```diff
@@ -0,0 +1,131 @@
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+	PDescribe("Run", func() {
+	})
```
why it is required?

jberkhahn replied:
I put in pending tests just as a reminder that they need to be implemented, but was going to wait until after kubebuilder is merged in because it'll require some additional refactoring. I can remove them if you think that's not needed.

camilamacedo86 replied:
could you please add // todo(jberkhahn):  {desc of what should be done} or remove them ?

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+	PDescribe("Run", func() {
+	})
```
Why it is required?

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd.go:
```diff
@@ -20,13 +20,13 @@ import (
+	DefaultChannel string
 	channels       string
-	generateOnly   bool
+	GenerateOnly   bool
```
why we need these changes?

jberkhahn replied:
I had to make a bunch of fields public to be able to set them in the tests. It's because cobra normally populated the fields set via flag when you initialize the command struct, but I can't do it that way because I'm not cobra. So I have to make them public so I can manually set them.

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd.go:
```diff
@@ -50,16 +50,16 @@ https://sdk.operatorframework.io/docs/olm-integration/
-		newCreateCmd(),
-		newValidateCmd(),
+		NewCreateCmd(),
+		NewValidateCmd(),
```
Why we need to export the NewCreateCmd and NewValidateCmd?

jberkhahn replied:
So i can call them from the tests.

camilamacedo86 commented on cmd/operator-sdk/bundle/cmd_test.go:
```diff
@@ -0,0 +1,65 @@
+	"github.com/operator-framework/operator-sdk/cmd/operator-sdk/bundle"
+)
+
+const use = "bundle"
+const short = "Manage operator bundle metadata"
+const long = `Manage bundle builds, bundle metadata generation, and bundle validation.
+An operator bundle is a portable operator packaging format understood by Kubernetes
+native software, like the Operator Lifecycle Manager.
+
+More information about operator bundles and metadata:
+https://github.com/operator-framework/operator-registry#manifest-format.
+
+More information about the integration with OLM via SDK:
+https://sdk.operatorframework.io/docs/olm-integration/
```
@estroz do you think that we should check if the cmd help string is ==?

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+	"github.com/operator-framework/operator-sdk/cmd/operator-sdk/bundle"
+)
+
+var _ = Describe("Running a bundle validate command", func() {
```
```suggestion
var _ = Describe("Checking operator-sdk bundle create command", func() {
```

jberkhahn replied:
ah, missed that. thanks

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+		})
+	})
+
+	Describe("Validate", func() {
```
Why the validate is in the create_test.go and not in the validate_test.go?

jberkhahn replied:
this is testing the argument validation method of the create command, not the bundle validate command

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+)
+
+var _ = Describe("Running a bundle validate command", func() {
+	Describe("NewCreateCmd", func() {
```
```suggestion
	Describe("NewCreateCmd should", func() {
```

jberkhahn replied:
I prefer to leave those describes as just the name of the method they're testing

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+			Expect(cmd.Use).To(Equal("create"))
+			Expect(cmd.Short).To(Equal("Create an operator bundle image"))
+			Expect(cmd.Long).To(Equal(`The 'operator-sdk bundle create' command will build an operator
+bundle image containing operator metadata and manifests, tagged with the
```
@estroz do you think that we should check if the cmd help string is ==? 
IMO I think we should test the command by verifying that it works as expected instead of it. 
Example:
operator-sdk bundle create
- Should successfully work when the flag --generated-only is informed and then should create in the disk a bundle with the files X. and then, we should ensure that the expected files were generated successfully. 
- Should fail when the crds are not found 
- Should fail when is not executed from the root path

jberkhahn replied:
these are intended to just unit test the cmd layer, not the entire command. So, testing that args are parsed and help text is output correctly, not to test the entire e2e functionality of the command, such as actually generating the output and writing to disk. Once KB integration has happened, I'd like to refactor the interface between the cmd parts and the pkg libs that actually do the stuff, and then add unit tests that ensure the correct methods are called with the right info, which is what the pending Run tests up above will be.

estroz replied:
Personally I don't see any benefit from comparing actual vs. expected command help strings. With these comparisons we'd have to update help text in 2 places. Perhaps `strings.TrimSpace(cmd.Long) != ""` but not the entire text.

jberkhahn replied:
Yeah, I was just trying to be complete. If I had TDD'd this package from the start, the tests would be the definitive source of the content, so I was just trying to replicate that. I can omit them if you think that's too cumbersome.

camilamacedo86 replied:
IMO we do not need to these checks at all. See that the tests added here did not change the % of coverage and are not testing the commands. Just if they have usage, long desc and etc.

jberkhahn replied:
These tests are covering the lines that set the long, usage, etc. for the command. I strongly think we should have a test that tests that the fields were populated correctly, but am open to suggestions for something more clever than Equal(). I'm not sure what estroz meant with the TrimSpace thing, doesn't that just trim leading and trailing whitespace?

camilamacedo86 replied:
Really tks for all your help closing it assorted out since we spoke today and agreed in the format to move forward.

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,221 @@
+			flag = cmd.Flags().Lookup("default-channel")
+			Expect(flag).NotTo(BeNil())
+			Expect(flag.Shorthand).To(Equal("e"))
+			Expect(flag.Usage).To(Equal("The default channel for the bundle image"))
```
IMO it is missing test/check when the commands should work and not.

jberkhahn replied:
I'm not sure I follow. Do you mean when the proper args are provided? That's what the Validate tests should be testing. Do you mean if all the proper files are present on disk that the generators need to run? I would argue that's the purview of the pkg libs that actually execute that, and should be covered by their unit tests.

camilamacedo86 replied:
The tests here are just checking the cli texts/usage when in POV should ensure the expected behaviour. Example:

- Should successfully work when the flag --generated-only is informed and then should create in the disk a bundle with the files X. and then, we should ensure that the expected files were generated successfully.
- Should fail when the crds are not found
- Should fail when is not executed from the root path

jberkhahn replied:
Yeah, that stuff is happening in the backend pkg libs which I currently don't have a way to fake out. The pending Run test would eventually test that, but we were going to put that refactor off until after the KB integration is finished.

camilamacedo86 replied:
Really tks for all your help closing it assorted out since we spoke today and agreed in the format to move forward.

estroz commented on cmd/operator-sdk/bundle/cmd_test.go:
```diff
@@ -0,0 +1,65 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bundle_test
```
Are you unable to write these tests in the same package? I've done this before without issues.
```suggestion
package bundle
```

jberkhahn replied:
You can, but I didn't on purpose. I'm trying to test the functionality of the package as a unit, so I don't want to be able to reach inside and use private stuff, as other packages won't be able to do that.

estroz replied:
I would agree if these packages were intended to be libraries. For now can you keep tests in the same package so we don't have to export a bunch of stuff? Before exporting private stuff we should have a conversation about general CLI refactoring.

jberkhahn replied:
ok, I'll move everything into a single package

jberkhahn commented:
updated. hopefully I didn't screw up squashing the commits

camilamacedo86 commented on cmd/operator-sdk/bundle/create.go:
```diff
@@ -32,7 +32,7 @@ import (
 )
 
-type bundleCreateCmd struct {
+type createCmd struct {
```
I agree that the name is redundant and we could/should change but it is not part of the scope of this PR. 
Could we do a PR just to do the renames?

jberkhahn replied:
woops, forgot to revert this when I moved everything back to private. fixed.

camilamacedo86 commented on cmd/operator-sdk/bundle/bundle_suite_test.go:
```diff
@@ -0,0 +1,27 @@
+// Copyright 2020 The Operator-SDK Authors
```
See that the tested added here did not increase the % of coverage:

<img width="763" alt="Screenshot 2020-06-20 at 09 43 30" src="https://user-images.githubusercontent.com/7708031/85197688-8b6ccf00-b2da-11ea-935d-f53fcbba15fa.png">

Also, see that what we really need to cover which are the funcs/cmd logic. See: https://coveralls.io/builds/31574282/source?filename=cmd/operator-sdk/bundle/create.go

In this PR we are checking if the cmd has long and short usages texts as if the default value set in the flag is there which is not required at all.  If we set the usage text and/or a default value using cobra it will be done by the module/cobra code source and not SDK. So, these behaviours are covered/tested by cobra in its repo and we do not need to ensure that. See: https://github.com/spf13/cobra/blob/master/doc/cmd_test.go 

Examples to help you with:


- [bundle/internal/result_test.go](https://github.com/operator-framework/operator-sdk/blob/master/cmd/operator-sdk/bundle/internal/result_test.go) 
- [blog -  How to test CLI commands made with Go and Cobra](https://gianarb.it/blog/golang-mockmania-cli-command-with-cobra) 

IMO: What we need to test?

We need call/execute the command passing the args and then check the expected results. 

Example:

`operator-sdk bundle create`

- Should successfully work when the flag --generated-only is informed and create a bundle in the disk
- Should fail when the crds are not found
- Should fail when is not executed from the root path

PS.: * See [here](https://github.com/operator-framework/operator-sdk/blob/master/internal/generate/olm-catalog/package_manifest_test.go) tests where we create tmp dirs. Note that any tmp data generated with the test need to be cleanup if it faces an issue and or an error.

jberkhahn replied:
I've already said that I intend to test Run() where the cmd actually does stuff, but am waiting on the kubebuilder integration as it will require a refactor. The test you linked is more like an e2e test as it's testing the actual execution of the cobra command, these are just unit tests testing the cmd layer alone.

camilamacedo86 replied:
Hi @jberkhahn, 

Just to clarifies: IMO what was added here is NOT required to be done/tested at all and I'd expected the cmd behaviours to be covered as described in the [blog -  How to test CLI commands made with Go and Cobra](https://gianarb.it/blog/golang-mockmania-cli-command-with-cobra).

camilamacedo86 replied:
Really tks for all your help closing it assorted out since we spoke today and agreed in the format to move forward.

jberkhahn commented:
Updated. Cut out most of the simple reflective tests, had to do some actual refactoring because something touched validate and added a new image builder, so I incorporated it into the validate func there and added a test for it.

camilamacedo86 commented on cmd/operator-sdk/bundle/validate.go:
```diff
@@ -126,7 +126,15 @@ func (c bundleValidateCmd) validate(args []string) error {
+
+	}
+
+	if c.imageBuilder != containertools.DockerTool.String() &&
+		c.imageBuilder != containertools.PodmanTool.String() &&
+		c.imageBuilder != containertools.NoneTool.String() {
+		return fmt.Errorf("unrecognized image-builder option: %s", c.imageBuilder)
 	}
+
 	return nil
```
It should be another PR it is not part of the ad unit test scope.

jberkhahn commented:
@camilamacedo86 updated

jberkhahn commented:
Is there anything else that needs to be changed here or is it good to go? @camilamacedo86 @estroz

camilamacedo86 commented on cmd/operator-sdk/bundle/create_test.go:
```diff
@@ -0,0 +1,145 @@
+		})
+	})
+
+	Describe("validate", func() {
```
Hi @jberkhahn, @estroz   

I do not think that the validate should be in the `create_test.go` since if we follow up the standard(go common practices) we will have the tests from create.go in the create_test.go and the test of validate.go in the validate_test.go.

Any reason for we do not follow up the common practice in this case?
